### PR TITLE
Let QPaintEvent tell us what region to repaint.

### PR DESCRIPTION
**Pull request by anntzer**
(instead of just updating the blitboxes.)

Alternate approach for #8948.
attn @tacaswell 

As for the performance questions, my work on mpl_cairo suggests to me that a big cost actually comes from the fact that we are copying the whole buffer into a string (https://github.com/matplotlib/matplotlib/blob/master/src/_backend_agg_wrapper.cpp#L586 before my recent changes, https://github.com/matplotlib/matplotlib/blob/master/src/_backend_agg_wrapper.cpp#L79 now that we always rely on Regions).  Note that the old Py2 version used to create a buffer object, which would avoid the need for copying.
In mpl_cairo, I instead return a numpy array which is a view on the internal buffer (https://github.com/anntzer/mpl_cairo/blob/master/src/_mpl_cairo.cpp#L260) and pass the address of the pointer (which can be retrieve via numpy) into QImage (https://github.com/anntzer/mpl_cairo/blob/master/mpl_cairo/qt.py#L15).  (To be clear, I haven't tried with agg, but on mpl_cairo this is much faster than copying the string around.)
This approach also removes the need for a refcount hack on PySide, as there is no buffer object for PySide to hold on anymore.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/whats_new.rst if major new feature
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 1]
```diff
@@ -62,15 +68,16 @@ def paintEvent(self, e):
...
-        if self._bbox_queue:
+        with self._bbox_queue_lock:
             bbox_queue = self._bbox_queue
-        else:
-            painter.eraseRect(self.rect())
-            bbox_queue = [
-                Bbox([[0, 0], [self.renderer.width, self.renderer.height]])]
+            self._bbox_queue = []
+
         self._bbox_queue = []
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5agg.py]: Don't need this any more

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 2]
```diff
@@ -62,15 +68,16 @@ def paintEvent(self, e):
...
         for bbox in bbox_queue:
+            if bbox is None:
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5agg.py]: If _any_ of them are `None` can short circuit the whole thing and just copy once.

**Comment by anntzer** [comment]
Actually I just rewrote the entire thing to use QPaintEvent.rect() (the idea you mentioned in #7366), which is much simpler.

**Comment by tacaswell** [comment]
Does `update` eventually produce a paint event with the full canvas as the event?

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 3]
```diff
@@ -134,7 +136,7 @@ def blit(self, bbox=None):
         """
         # If bbox is None, blit the entire canvas. Otherwise
         # blit only the area defined by the bbox.
-        if bbox is None and self.figure:
+        if bbox is None:
             bbox = self.figure.bbox
 
         self._bbox_queue.append(bbox)
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5agg.py]: might as well drop this line too?

**Comment by anntzer** [comment]
It does: http://doc.qt.io/qt-4.8/qwidget.html#update

File: lib/matplotlib/backends/backend_qt5agg.py (hunk 3 above)
Comment (anntzer) [inline lib/matplotlib/backends/backend_qt5agg.py]: That would technically be backwards incompatible (and the behavior is "mandated" by the docstring of `FigureCanvasBase`), but a quick grep suggests that we never rely on it (always explicitly passing in the figure bbox when needed) so I'm fine with getting rid of the behavior.  You get to decide on the deprecation policy :-)

File: lib/matplotlib/backends/backend_qt5agg.py [hunk 4]
```diff
@@ -60,32 +61,32 @@ def paintEvent(self, e):
...
+        left = rect.left()
+        top = rect.top()
+        width = rect.width()
+        height = rect.height()
+        if self._erase_before_paint:
             painter.eraseRect(self.rect())
-            bbox_queue = [
-                Bbox([[0, 0], [self.renderer.width, self.renderer.height]])]
-        self._bbox_queue = []
-        for bbox in bbox_queue:
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_qt5agg.py]: One concern is that Qt drops some paint events, but that would be surprising.

Comment (anntzer) [inline lib/matplotlib/backends/backend_qt5agg.py]: If Qt pulls the rug under us, we're cooked anyways, no?

**Comment by anntzer** [comment]
rebased

**Pull request merged**